### ISSUE TITLE
Broken entity_hidden type when using custom identifier

### ISSUE THREAD

**Issue by waaghals** [Sylius/Sylius#2331]
The `entity_hidden` type seems broken.
I have a custom variant with a bar code property.

When configuring the an `entity_hidden` type for this property it fails in the `ObjectToIdentifierTransformer` on the third pass.

1. `ObjectToIdentifierTransformer::transform()` gets called via the `form.pre_bind` (pre submit) event and a string with the bar code is passed. This run  is successful and returns the correct entity;
2. `ObjectToIdentifierTransformer::transform()` gets called via the `form.bind` (submit) event and again a bar code is passed. This second pass returns the correct entity as well.
3. On the third pass to `ObjectToIdentifierTransformer::transform()` gets called by the form during the view transformation. But this time the argument passed to the the `transform()` method is already a valid entity. 

When the identifier is set to the id, it works fine. But when I use a other property than id it fails when finding the entity from the repository:

In  `ObjectToIdentifierTransformer::transform()`:
```php
if (null === $entity = $this->repository->findOneBy(array($this->identifier => $value))) {
    throw new TransformationFailedException(sprintf(
        'Object "%s" with identifier "%s"="%s" does not exist.',
        $this->repository->getClassName(),
        $this->identifier,
        $value
    ));
}
```

Why is there a viewTransformer set when the values are updated using the form events?
I modified the `ObjectToIdentifierTransformer::transform()` method to the following.
This skips the repository when the passed `$value` is already in the desired form.
But this feels like a hack because imho the transform method shouldn't even be called with the entity data. Only with the property value for which you want to find the entity.
```php
   public function transform($value)
    {
        if (!$value) {
            return null;
        }

        $className = $this->repository->getClassName();
        if(is_object($value) && $value instanceof $className) {
            return $value;
        }

        if (null === $entity = $this->repository->findOneBy(array($this->identifier => $value))) {
            throw new TransformationFailedException(sprintf(
                'Object "%s" with identifier "%s"="%s" does not exist.',
                $this->repository->getClassName(),
                $this->identifier,
                $value
            ));
        }

        return $entity;
    }
```

*Sidenote*:
The `ObjectToIdentifierTransformer` is hard-coded in the `entity_type`. Now I have to extend the `entity_type` just so I can replace the  `ObjectToIdentifierTransformer`.
```php
class EntityHiddenType extends AbstractType
{
    ...
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        ...
        $transformer = new ObjectToIdentifierTransformer($this->manager->getRepository($options['data_class']), $options['identifier']);

        $builder->...
    }
    ...
}
```

**Comment by adamelso** [comment]
I was having a similar issue previously, where I created a PR #2415 that I thought would fix the problem. 

It turns out that `entity_hidden` will only work if its only added to the form builder on `PRE_SUBMIT`.

Example (which is PHP 5.5 code):

```php
->addEventListener(FormEvents::PRE_SUBMIT, function (FormEvent $event) use ($options) {
    $data = $event->getData();
    if (isset($data['variant'])) {
        $event->getForm()->add('variant', 'entity_hidden', [
            'data_class' => ProductVariant::class,
            'identifier' => 'barCode',
        ]);
    }
})
```

Can you confirm that this also works when using a custom identifier other than the ID?

**Comment by pjedrzejewski** [comment]
``entity_hidden`` will be removed and replaced with nicer generic field for resources.

**Comment by michalmarcinkowski** [comment]
Should be fixed in #3831
